packages/ref/src/cmip_ref/executor/__init__.py

```diff
@@ -58,3 +66,76 @@ def import_executor_cls(fqn: str) -> type[Executor]:
     except AttributeError:
         logger.error(f"Provider '{fqn}' not found")
         raise InvalidExecutorException(fqn, f"Executor '{attribute_name}' not found in {module}")
+
+
+def _copy_file_to_results(
+    scratch_directory: pathlib.Path,
+    results_directory: pathlib.Path,
+    fragment: pathlib.Path | str,
+    filename: pathlib.Path | str,
+) -> None:
+    """
+    Copy a file from the scratch directory to the results directory
+
+    Parameters
+    ----------
+    scratch_directory
+        The directory where the file is currently located
+    results_directory
+        The directory where the file should be copied to
+    fragment
+        The fragment of the results directory where the file should be copied
+    filename
+        The name of the file to be copied
+    """
+    output_directory = results_directory / fragment
+
+    if not (scratch_directory / fragment / filename).exists():
+        raise FileNotFoundError(f"Could not find {filename} in {scratch_directory / fragment}")
+
+    if not output_directory.exists():
+        output_directory.mkdir(parents=True)
+
+    shutil.copy(scratch_directory / fragment / filename, output_directory / filename)
+
+
+def handle_execution_result(
+    config: "Config", metric_execution_result: "MetricExecutionResult", result: "MetricResult"
+) -> None:
+    """
+    Handle the result of a metric execution
+
+    This will update the metric execution result with the output of the metric execution.
+    The output will be copied from the scratch directory to the results directory.
+
+    Parameters
+    ----------
+    config
+        The configuration to use
+    metric_execution_result
+        The metric execution result to update
+    result
+        The result of the metric execution, either successful or failed
+    """
+    if result.successful and result.bundle_filename is not None:
+        logger.info(f"{metric_execution_result} successful")
+
+        # TODO: Iterate over the files in the bundle and copy them
+        # TODO: Add files to db
+        # Rework after #99
+        _copy_file_to_results(
+            config.paths.scratch,
+            config.paths.results,
+            metric_execution_result.output_fragment,
+            result.bundle_filename,
+        )
+
+        metric_execution_result.mark_successful(result.bundle_filename)
+
+        # TODO: This should check if the result is the most recent for the execution,
+        # if so then update the dirty fields
+        # i.e. if there are outstanding results don't make as clean
+        metric_execution_result.metric_execution.dirty = False
```

Comment:
In an async world where new data comes in while we are processing the old data, handling this all via dirty fields by hand sounds harsh.

Reply:
Yup I'm not super happy with this approach. The problem I'm trying to solve is the option to mark a execution as dirty to trigger it to be rerun on the next solve. It should stay dirty until a run started **after** being marked as dirty is complete.

Reply:
maybe we just need dirty timestamps instead of booleans? "Dirty since"?

Reply:
Or even dirty uuids?

Reply:
So the dirty state would be cleared if the created datetime of a reslt is after the dirty datetime. I'm not quite sure if I follow the uuid thought. Probably worth an issue as it might be a good hackathon task

Reply:
the uuid would be that when an execution starts, ideally atomically at the point when it collects input data, it also collects the dirty uuids that it will clean, and once we have the result, only if dirty uuid is still the same, it is cleared.